### ISSUE TITLE
abort because of "non-virtual thunk to grpc::Channel::PerformOpsOnCall(...) channel_cc.cc:130"

### ISSUE THREAD

**Issue by xqzhang2015** [grpc/grpc#20754]
#19736 # What version of gRPC and what language are you using?
gRPC version: 1.14.0
language: C++

### What operating system (Linux, Windows,...) and version?

```sh
!30 $ uname -a
Linux xxx 3.10.0-327.13.1.el7.x86_64 #1 SMP Thu Mar 31 16:04:38 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
!32 $ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
```

### What runtime / compiler are you using (e.g. python version or version of gcc)
```sh
clang version 3.8.1 (tags/RELEASE_381/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/local/bin
Found candidate GCC installation: /usr/lib/gcc/i686-redhat-linux/4.8.2
```

### What did you do?
I failed to reproduce the issue in dev env. It only happened 2 times in stg env.

I'm using gRPC to develop a C++version etcd client. It happened at write request on a stream for etcd lease.

### What did you expect to see?
gRPC doesn't abort.

### What did you see instead?
[stack trace]

```sh
#0  0x00007f58a94d15f7 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install glibc-2.17-106.el7_2.4.x86_64 keyutils-libs-1.5.8-3.el7.x86_64 krb5-libs-1.13.2-12.el7_2.x86_64 libcom_err-1.42.9-7.el7.x86_64 libgcc-4.8.5-4.el7.x86_64 libselinux-2.2.2-6.el7.x86_64 libstdc++-4.8.5-4.el7.x86_64 libtool-ltdl-2.4.2-21.el7_2.x86_64 libxml2-2.9.1-6.el7_2.2.x86_64 libxslt-1.1.28-5.el7.x86_64 nss-softokn-freebl-3.16.2.3-14.2.el7_2.x86_64 openssl-libs-1.0.1e-51.el7_2.4.x86_64 pcre-8.32-15.el7.x86_64 unixODBC-2.3.1-11.el7.x86_64 xerces-c-3.1.1-8.el7_2.x86_64 xz-libs-5.1.2-12alpha.el7.x86_64 zlib-1.2.7-15.el7.x86_64
(gdb)
(gdb) bt
#0  0x00007f58a94d15f7 in raise () from /lib64/libc.so.6
#1  0x00007f58a94d2ce8 in abort () from /lib64/libc.so.6
#2  0x0000000001f14100 in non-virtual thunk to grpc::Channel::PerformOpsOnCall(grpc::internal::CallOpSetInterface*, grpc::internal::Call*) () at src/cpp/client/channel_cc.cc:130
#3  0x0000000001e5d8d8 in PerformOps (this=<optimized out>, ops=<optimized out>)
    at /workspace/common/src/3rd/grpc/grpc-1.14.0/include/grpcpp/impl/codegen/call.h:680
#4  grpc::ClientAsyncReaderWriter<etcdserverpb::LeaseKeepAliveRequest, etcdserverpb::LeaseKeepAliveResponse>::Write (this=0x7f5792cfff20, msg=..., tag=<optimized out>)
    at /workspace/common/src/3rd/grpc/grpc-1.14.0/include/grpcpp/impl/codegen/async_stream.h:536
#5  0x0000000001e5c3d7 in etcd::LeaseImpl::write_next_lease_keep_alive (this=0x7f58a73568c0)
    at /workspace/common/src/3rd/etcdv3-cpp/src/lease_impl.cpp:561
...
```

### Anything else we should know about your project / environment?
The abort is at GPR_ASSERT. There is no local return value for grpc_call_start_batch(). So I don't where the issue is.

```c++
void Channel::PerformOpsOnCall(internal::CallOpSetInterface* ops,
                               internal::Call* call) {
  static const size_t MAX_OPS = 8;
  size_t nops = 0;
  grpc_op cops[MAX_OPS];
  ops->FillOps(call->call(), cops, &nops);
  GPR_ASSERT(GRPC_CALL_OK ==
             grpc_call_start_batch(call->call(), cops, nops, ops, nullptr));
}
```

**Comment by vjpai** [comment]
This assertion indicates an API misuse of some kind. I have a PR in flight to provide some additional info on this log, but the most common triggers for this are to have multiple writes outstanding at the same time, or to have some object dead when it needs to be alive (e.g., ClientContext needs to be alive for the entire duration of the RPC). Can you provide any additional information about your use of the code? Since you mentioned write request on a stream and I see that in the trace: are you following the 1-outstanding-write rule?

**Comment by xqzhang2015** [comment]
@vjpai Thanks for your quick reply.

I know the 1-outstanding-write rule but not 100% sure if my code complies with the rule.

For this etcd c++ client, I use state-machine to loop keep lease alive and handling all kinds of exceptions, including network issues, which is a little complicated. Let me review it first.

Is it possible that the issue fixed in 1.14.1 will cause such aborting(or crash)?
- https://github.com/grpc/grpc/releases/tag/v1.14.1

BTW, I got one issue similar to mine and the corresponding fix. Is that a gRPC lib issue? or gRPC usage issue?
- https://github.com/firebase/firebase-ios-sdk/issues/2138
- https://github.com/firebase/firebase-ios-sdk/pull/2146/files
